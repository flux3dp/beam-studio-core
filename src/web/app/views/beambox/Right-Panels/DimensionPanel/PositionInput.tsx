import React, { memo, useCallback, useMemo } from 'react';

import ObjectPanelItem from 'app/views/beambox/Right-Panels/ObjectPanelItem';
import UnitInput from 'app/widgets/Unit-Input-v2';
import storage from 'implementations/storage';
import { useIsMobile } from 'helpers/system-helper';

import styles from './DimensionPanel.module.scss';

interface Props {
  type: 'x' | 'y' | 'x1' | 'y1' | 'x2' | 'y2' | 'cx' | 'cy';
  value: number;
  onChange: (type: string, value: number) => void;
}

const PositionInput = ({ type, value, onChange }: Props): JSX.Element => {
  const isMobile = useIsMobile();
  const unit = useMemo(() => (storage.get('default-units') === 'inches' ? 'in' : 'mm'), []);
  const label = useMemo<string | JSX.Element>(() => {
    if (type === 'x') return 'X';
    if (type === 'y') return 'Y';
    if (type === 'x1')
      return (
        <>
          X<sub>1</sub>
        </>
      );
    if (type === 'y1')
      return (
        <>
          Y<sub>1</sub>
        </>
      );
    if (type === 'x2')
      return (
        <>
          X<sub>2</sub>
        </>
      );
    if (type === 'y2')
      return (
        <>
          Y<sub>2</sub>
        </>
      );
    if (type === 'cx')
      return (
        <>
          X<sub>C</sub>
        </>
      );
    if (type === 'cy')
      return (
        <>
          Y<sub>C</sub>
        </>
      );
    return null;
  }, [type]);
  const inputId = useMemo(() => `${type}_position`, [type]);
  const handleChange = useCallback((val: number) => onChange(type, val), [type, onChange]);
  if (isMobile)
    return (
      <ObjectPanelItem.Number id={inputId} value={value} updateValue={handleChange} label={label} />
    );
  return (
    <div className={styles.dimension}>
      <div className={styles.label}>{label}</div>
      <UnitInput id={inputId} unit={unit} defaultValue={value} getValue={handleChange} />
    </div>
  );
};

export default memo(PositionInput);
