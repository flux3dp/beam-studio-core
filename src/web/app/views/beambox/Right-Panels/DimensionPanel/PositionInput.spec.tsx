import React from 'react';
import { fireEvent, render } from '@testing-library/react';

import PositionInput from './PositionInput';

const mockGet = jest.fn();
jest.mock('implementations/storage', () => ({
  get: (...args) => mockGet(...args),
}));

const mockUseIsMobile = jest.fn();
jest.mock('helpers/system-helper', () => ({
  useIsMobile: () => mockUseIsMobile(),
}));

jest.mock('app/widgets/Unit-Input-v2', () => ({ id, unit, defaultValue, getValue }: any) => (
  <div>
    <div>{id}</div>
    <div>{unit}</div>
    <div>{defaultValue}</div>
    <div>{getValue}</div>
    <input type="number" defaultValue={defaultValue} onChange={(e) => getValue(Number(e.target.value))} />
  </div>
));

jest.mock('app/views/beambox/Right-Panels/ObjectPanelItem', () => ({
  Number: ({ id, value, updateValue, label }: any) => (
    <div id={id}>
      {label}
      <div>{value}</div>
      <button type="button" onClick={() => updateValue(value + 1)} />
    </div>
  ),
}));

const mockOnChange = jest.fn();

describe('test PositionInput', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('should render correctly on desktop', () => {
    mockUseIsMobile.mockReturnValue(false);
    const { container } = render(<PositionInput type="x" value={0} onChange={mockOnChange} />);
    expect(container).toMatchSnapshot();
  });

  it('should render correctly on mobile', () => {
    mockUseIsMobile.mockReturnValue(true);
    const { container } = render(<PositionInput type="x" value={0} onChange={mockOnChange} />);
    expect(container).toMatchSnapshot();
  });

  test('onChange on desktop', () => {
    mockUseIsMobile.mockReturnValue(false);
    const { container } = render(<PositionInput type="x" value={0} onChange={mockOnChange} />);
    const input = container.querySelector('input');
    fireEvent.change(input, { target: { value: 1 } });
    expect(mockOnChange).toBeCalledTimes(1);
    expect(mockOnChange).toHaveBeenLastCalledWith('x', 1);
  });

  test('onChange on mobile', () => {
    mockUseIsMobile.mockReturnValue(true);
    const { container } = render(<PositionInput type="x" value={0} onChange={mockOnChange} />);
    const button = container.querySelector('button');
    fireEvent.click(button);
    expect(mockOnChange).toBeCalledTimes(1);
    expect(mockOnChange).toHaveBeenLastCalledWith('x', 1);
  });
});
