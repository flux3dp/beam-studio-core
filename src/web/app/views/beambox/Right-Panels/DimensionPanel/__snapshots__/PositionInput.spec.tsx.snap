// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test PositionInput should render correctly on desktop 1`] = `
<div>
  <div
    class="dimension"
  >
    <div
      class="label"
    >
      X
    </div>
    <div>
      <div>
        x_position
      </div>
      <div>
        mm
      </div>
      <div>
        0
      </div>
      <div />
      <input
        type="number"
        value="0"
      />
    </div>
  </div>
</div>
`;

exports[`test PositionInput should render correctly on mobile 1`] = `
<div>
  <div
    id="x_position"
  >
    X
    <div>
      0
    </div>
    <button
      type="button"
    />
  </div>
</div>
`;
